### PR TITLE
Add renewal support for AT and internal registrations in DomainCreateFlow

### PR DESCRIPTION
**Pull request by rachelguan**
Renewal info will be added to recurring billing event when creating a domain (create flow). The code change mostly happens in `run()` which contains the logic of the flow. Since renewal price behavior needs to be set and that could be determined using existing boolean `isAnchorTenant` and `isValidReservedCreate`, a new helper method `getRenewalPriceBehavior` was created in this PR to determine the state of renewal price behavior, and this state replace the two existing booleans. As I was going through the flow, I noticed that some helper methods could be adjusted for minor improvements. For example, `isSunriseCreate` and `isReserved(domainName, isSunriseCreate) `were passed into `createOneTimeBillingEvent` and it could be `domainName` and `isSunriseCreate` and put the conditional in the helper method instead. 

tl;dr: this PR does the following: 
- Set the info for recurring billing event including renewal price behavior and renewal price 
- Replaced existing boolean `isAnchorTenant` and `isReserved` with `RenewalPriceBehavior.NONPREMIUM` and `RenewalPriceBehavior.SPECIFIED` 
- Modified existing testing set up to accommodate the renewal feature for AT and internal registrations.
- Rearranged parameters of some of the helper methods 

---
This change is [Reviewable](https://reviewable.io/reviews/google/nomulus/1591)

### CODE REVIEW

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1591)* status: 0 of 2 files reviewed, 8 unresolved discussions (waiting on @CydeWeys and @rachelguan)

___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 312 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06mnTr4ZZOJVXGQ_qB:-N06k34j8JG985hPiczA:b-juwjr8) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L312)):*
> ```Java
>             registry, targetId, now, years, isAnchorTenant, allocationToken);
>     RenewalPriceInfo renewalPriceInfo =
>         getRenewalPriceInfo(isAnchorTenant, allocationToken, feesAndCredits.getCreateCost());
> ```

It's minor, but I think this is clearer if you simply pass in `feesAndCredits` to the helper function, and then grab the create cost from within there.  Otherwise it kind of looks like an error at the callsite, as in, why are we calling `getCreateCost()` instead of `getRenewCost()` when calculating coming called the `renewalPriceInfo`?
___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 633 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06klrx30smLg1HTa1c:-N06klrx30smLg1HTa1d:bmewsd6) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L633)):*
> ```Java
>    */
>   static RenewalPriceInfo getRenewalPriceInfo(
>       Boolean isAnchorTenant, Optional<AllocationToken> allocationToken, Money createCost) {
> ```

Should be `boolean`, not `Boolean`.  Only use boxed primitives when necessary.
___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 634 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06kqIE9jejizytWYgB:-N06kqIE9jejizytWYgC:bkpuddi) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L634)):*
> ```Java
>   static RenewalPriceInfo getRenewalPriceInfo(
>       Boolean isAnchorTenant, Optional<AllocationToken> allocationToken, Money createCost) {
>     checkNotNull(createCost, "Create cost cannot be null");
> ```

`checkArgumentNotNull()`  throws a better exception type.
___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 635 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06l8cp3dxO93xmWXbe:-N06l8cp3dxO93xmWXbf:b-cxr9n0) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L635)):*
> ```Java
>       Boolean isAnchorTenant, Optional<AllocationToken> allocationToken, Money createCost) {
>     checkNotNull(createCost, "Create cost cannot be null");
>     if (isAnchorTenant) {
> ```

I think the overall logic is clearer here if it looks like this:

```
switch renewal price behavior:
  case NONPREMIUM:
    special handling for anchor tenants
  case SPECIFIED:
    ...
  case DEFAULT:
    ...
  default:
    throw Exception
```

As written, your `else` will cause the wrong thing to happen if another type of renewal price behavior is added, whereas my default case will make it very clear that the user failed to add the correct handling here for that.  In general you should be using explicit handling for all current values of an enum and then throwing if an unexpected value comes in.
___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2617 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nLt1E3ryXJ8d8qjA:-N06nLt1E3ryXJ8d8qjB:b-h1phb2) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2617)):*
> ```Java
>       throws EppException {
>     assertThat(
>             DomainCreateFlow.getRenewalPriceInfo(
> ```

If you statically import this you won't need as much line-wrapping, and there'll be no loss of clarity.
___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2663 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nf7v0-zc4I3kalbK:-N06nf7v0-zc4I3kalbL:b-z5urq2) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2663)):*
> ```Java
>                 .build());
>     assertThat(
>             assertThrows(
> ```

We don't typically wrap these two assertions like this in the rest of our codebase and it may not be the best idea?  Unclear.
___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2665 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nIBXECkMmIuk_pd0:-N06nIBXECkMmIuk_pd1:b-dlss02) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2665)):*
> ```Java
>             assertThrows(
>                 IllegalArgumentException.class,
>                 () -> {
> ```

Unnecessary block
___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2679 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nGOW9PzyYthbpCse:-N06nGOW9PzyYthbpCsf:b-dlss02) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2679)):*
> ```Java
>             assertThrows(
>                 NullPointerException.class,
>                 () -> {
> ```

Unnecessary block

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1591)* status: 0 of 2 files reviewed, 8 unresolved discussions (waiting on @CydeWeys and @rachelguan)

___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 635 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06l8cp3dxO93xmWXbe:-N07reE6BMze4jJS1klO:b-76zrb4) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L635)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

I think the overall logic is clearer here if it looks like this:

```
switch renewal price behavior:
  case NONPREMIUM:
    special handling for anchor tenants
  case SPECIFIED:
    ...
  case DEFAULT:
    ...
  default:
    throw Exception
```

As written, your `else` will cause the wrong thing to happen if another type of renewal price behavior is added, whereas my default case will make it very clear that the user failed to add the correct handling here for that.  In general you should be using explicit handling for all current values of an enum and then throwing if an unexpected value comes in.
</blockquote></details>

To be clearer:

```
if allocation token is present:
  switch on allocation token's renewal price behavior:
    case NONPREMIUM:
      special handling for anchor tenants
    case SPECIFIED:
      ...
    case DEFAULT:
      ...
    default:
      throw Exception
else:
  do basic handling
```

**Review by rachelguan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1591)* status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @CydeWeys)

___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 312 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06mnTr4ZZOJVXGQ_qB:-N07pvBc3zpqFj2e9Ue_:b-896fix) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L312)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

It's minor, but I think this is clearer if you simply pass in `feesAndCredits` to the helper function, and then grab the create cost from within there.  Otherwise it kind of looks like an error at the callsite, as in, why are we calling `getCreateCost()` instead of `getRenewCost()` when calculating coming called the `renewalPriceInfo`?
</blockquote></details>

Done.
___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 633 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06klrx30smLg1HTa1c:-N07ptz_8IH5GSLBjgIR:b-896fix) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L633)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Should be `boolean`, not `Boolean`.  Only use boxed primitives when necessary.
</blockquote></details>

Done.
___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 634 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06kqIE9jejizytWYgB:-N07puTo1QZ0DLGbpEj9:b-896fix) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L634)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

`checkArgumentNotNull()`  throws a better exception type.
</blockquote></details>

Done.
___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 635 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06l8cp3dxO93xmWXbe:-N0CLssj8p0JUJ1ULUBb:brv3gl0) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L635)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

To be clearer:

```
if allocation token is present:
  switch on allocation token's renewal price behavior:
    case NONPREMIUM:
      special handling for anchor tenants
    case SPECIFIED:
      ...
    case DEFAULT:
      ...
    default:
      throw Exception
else:
  do basic handling
```
</blockquote></details>

I actually find it easier to read to use `isAnchorTenant` first to determine the anchor tenants since that only requires the boolean for one behavior..then its either specified (if there's token that says specified), or default..

It's nice to have a basic handling for exceptions in theory but there's already a check added in `setRenewalPriceBehaviort` in `AllocationToken` that it cant be `NONPREMIUM`.  

And now `case DEFAULT` has the same `if ..else` as  the case for the absence of allocation token, 
___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2663 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nf7v0-zc4I3kalbK:-N0DMPJ3CJXfSnN_Usm4:b-2qjwah) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2663)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

We don't typically wrap these two assertions like this in the rest of our codebase and it may not be the best idea?  Unclear.
</blockquote></details>

I've seen that in the code base and i fixed it here. 
___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2665 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nIBXECkMmIuk_pd0:-N0DCvt878ozubgERofx:b-t0c8if) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2665)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Unnecessary block
</blockquote></details>

do you mean that its ok to have `isAnchorTenant = true` and renewal price behavior = `SPECIFIED` should not be blocked? 
___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2679 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nGOW9PzyYthbpCse:-N0DD8SZAcpPtriKKOTX:b-9a4utc) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2679)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Unnecessary block
</blockquote></details>

do you mean it's ok if the create cost is null ?

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1591)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @CydeWeys and @rachelguan)

___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2665 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nIBXECkMmIuk_pd0:-N0HefL49U9KoNYPoxsJ:b-jablc5) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2665)):*
<details><summary><i>Previously, rachelguan (Rachel Guan) wrote…</i></summary><blockquote>

do you mean that its ok to have `isAnchorTenant = true` and renewal price behavior = `SPECIFIED` should not be blocked? 
</blockquote></details>

The curly braces are unnecessary.  You only need them when your lambda consists of multiple statements.  

Block: http://beginwithjava.blogspot.com/2008/06/code-blocks.html  (apologies for the introductory nature of the link but I couldn't find any good reference material to explain the nomenclature)
___
*[core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java, line 2679 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nGOW9PzyYthbpCse:-N0HemMB9O7XrVcjl52w:bm9le5y) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2679)):*
<details><summary><i>Previously, rachelguan (Rachel Guan) wrote…</i></summary><blockquote>

do you mean it's ok if the create cost is null ? 
</blockquote></details>

Curly braces again.

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1591)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @CydeWeys and @rachelguan)

___
*[core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java, line 635 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06l8cp3dxO93xmWXbe:-N0HhoUg1giF2uqoeJaf:bkbsywr) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L635)):*
<details><summary><i>Previously, rachelguan (Rachel Guan) wrote…</i></summary><blockquote>

I actually find it easier to read to use `isAnchorTenant` first to determine the anchor tenants since that only requires the boolean for one behavior..then its either specified (if there's token that says specified), or default..

It's nice to have a basic handling for exceptions in theory but there's already a check added in `setRenewalPriceBehaviort` in `AllocationToken` that it cant be `NONPREMIUM`.  

And now `case DEFAULT` has the same `if ..else` as  the case for the absence of allocation token, 
</blockquote></details>

Hrm, I guess you're right, that is actually easier, now that I see this.

**Review by rachelguan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1591)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @CydeWeys)

___
*[`core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java` line 635 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06l8cp3dxO93xmWXbe:-N0ba7QjErPsDmbeJFP_:bik7zgz) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/main/java/google/registry/flows/domain/DomainCreateFlow.java#L635)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Hrm, I guess you're right, that is actually easier, now that I see this.
</blockquote></details>

reverted it back to the implementation from last commit. 
___
*[`core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java` line 2665 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nIBXECkMmIuk_pd0:-N0baC8M2fKRNxs5RuAj:bk084gq) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2665)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

The curly braces are unnecessary.  You only need them when your lambda consists of multiple statements.  

Block: http://beginwithjava.blogspot.com/2008/06/code-blocks.html  (apologies for the introductory nature of the link but I couldn't find any good reference material to explain the nomenclature)
</blockquote></details>

misunderstood the previous comment. 
___
*[`core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java` line 2679 at r2](https://reviewable.io/reviews/google/nomulus/1591#-N06nGOW9PzyYthbpCse:-N0baNKW6VHTwIMdSBKL:bw3gryy) ([raw file](https://github.com/google/nomulus/blob/1192b00f9f7a0d4d5ffa500cbcf64a01baea0e01/core/src/test/java/google/registry/flows/domain/DomainCreateFlowTest.java#L2679)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Curly braces again.
</blockquote></details>

misunderstood and fixed.

**Review by CydeWeys — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/google/nomulus/1591#-:-N0lb6Pb3HE3HZE4zZwB:bnfp4nl)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @CydeWeys)
